### PR TITLE
Do not run getRepairSteps in register_commands

### PR DESCRIPTION
**Pull request by kesselb**
`getRepairSteps` is quite expensive (because every repair step is initialized and their dependencies are injected).

screenshot from 2019-02-02 18-39-51

How to reproduce:

1. Monkey patch [Repair.php](https://github.com/nextcloud/server/blob/ecc8068e66abe77785cccffa0e4da1b5c209ebca/lib/private/Repair.php#L135) and add `echo 'getRepairSteps has been called' . PHP_EOL;`

2. Run `occ app:list`

3. You see "getRepairSteps has been called"

### CODE REVIEW

**Review by MorrisJobke — APPROVED** [review]
Code makes sense and works 👍

File: lib/private/Repair.php [hunk 1]
```diff
@@ -73,9 +72,9 @@ class Repair implements IOutput {
 	 * Creates a new repair step runner
 	 *
 	 * @param IRepairStep[] $repairSteps array of RepairStep instances
-	 * @param EventDispatcher $dispatcher
+	 * @param EventDispatcherInterface $dispatcher
 	 */
-	public function __construct($repairSteps = [], EventDispatcher $dispatcher = null) {
+	public function __construct($repairSteps = [], EventDispatcherInterface $dispatcher = null) {
```
Comment (nickvergessen) [inline lib/private/Repair.php]: Could also remove the empty array now?
Also I guess it's not called without dependency injection anymore so EventDispatcherInterface can be made non-optional?

Comment (danielkesselberg) [reply lib/private/Repair.php]: <pre>$ grep -r -E &apos;new Repair\(|new \\OC\\Repair\(&apos;
<font color="#75507B">tests/lib/RepairStepTest.php</font><font color="#06989A">:</font>		$this-&gt;repair = <font color="#EF2929"><b>new \OC\Repair(</b></font>[], $dispatcher);
<font color="#75507B">lib/private/legacy/app.php</font><font color="#06989A">:</font>		$r = <font color="#EF2929"><b>new Repair(</b></font>[], $dispatcher);
<font color="#75507B">lib/private/Updater.php</font><font color="#06989A">:</font>		$repair = <font color="#EF2929"><b>new Repair(</b></font>Repair::getBeforeUpgradeRepairSteps(), \OC::$server-&gt;getEventDispatcher());
<font color="#75507B">lib/private/Updater.php</font><font color="#06989A">:</font>		$repair = <font color="#EF2929"><b>new Repair(</b></font>Repair::getRepairSteps(), \OC::$server-&gt;getEventDispatcher());
<font color="#75507B">lib/private/Migration/BackgroundRepair.php</font><font color="#06989A">:</font>		$repair = <font color="#EF2929"><b>new Repair(</b></font>[], $this-&gt;dispatcher);
<font color="#75507B">core/register_command.php</font><font color="#06989A">:</font>		<font color="#EF2929"><b>new \OC\Repair(</b></font>[], \OC::$server-&gt;getEventDispatcher()),
</pre>

I would say yes

**Comment by MorrisJobke** [comment]
Conflicts

**Comment by danielkesselberg** [comment]
> Conflicts

Rebased

**Comment by danielkesselberg** [comment]
https://github.com/nextcloud/server/pull/13995/commits/040295ae5410a9377f30659a2343c66f7117f4a1 fixed a failing unit test. `$dispatcher` was null for one test scenario. Changed signature for `setDispatcher`. Could not find any usage except this test.

**Review by rullzer — APPROVED** [review]

**Pull request merged**